frontend/src/routes/project/[project_code]/AddProjectUser.svelte

```diff
@@ -0,0 +1,68 @@
+﻿<script lang="ts">
+	import { _addProjectUser } from './+page';
+	import Form from '$lib/forms/Form.svelte';
+	import Input from '$lib/forms/Input.svelte';
+	import t from '$lib/i18n';
+	import Modal, { DialogResponse } from '$lib/components/modals/Modal.svelte';
+	import { ProjectRole } from '$lib/gql/graphql';
+	import { lexSuperForm, lexSuperValidate, Select } from '$lib/forms';
+	import { z } from 'zod';
+
+	const formSchema = z.object({
+		email: z.string().email($t('project_page.add_user.email_required')),
+		role: z.enum([ProjectRole.Editor, ProjectRole.Manager])
+	});
+
+	export let projectId: string;
+	let modal: Modal;
+	let {form, errors, valid, update, reset, message } = lexSuperForm(formSchema);
+	async function openModal(): Promise<void> {
+		if (await modal.openModal() === DialogResponse.Cancel) return;
+		//validate form, we're using a wrapper over the library method because we're doing this all client side.
+		await lexSuperValidate($form, formSchema, update);
+		// go back to the top and await a response from the modal again.
+		if (!$valid) return await openModal();
+		const result = await _addProjectUser({ projectId, userEmail: $form.email, role: $form.role });
+		if (result.error) {
+			$message = result.error.message;
+			// again go back to the top and await a response from the modal.
+			return await openModal();
+		}
+		modal.close();
+	}
+</script>
+
+<button class="badge badge-lg badge-success cursor-pointer" on:click={openModal}>
+	<span class="i-mdi-plus"/> {$t('project_page.add_user.add_button')}
+</button>
+<Modal bind:this={modal} on:close={() => reset()} bottom>
+	<Form on:submit={() => modal.submitModal()}>
+		<p>{$t('project_page.add_user.modal_title')}</p>
+		<Input
+			id="email"
+			type="email"
+			label={$t('admin_dashboard.column_email')}
+			bind:value={$form.email}
+			required
+			error={$errors.email}
+			autofocus
+		/>
+		<Select id="select-role" bind:value={$form.role} label={$t('project_role.label')} error={$errors.role}>
+			<option value={ProjectRole.Editor}>{$t('project_role.editor_description')}</option>
+			<option value={ProjectRole.Manager}>{$t('project_role.manager_description')}</option>
+		</Select>
+	</Form>
+	<svelte:fragment slot="actions" let:closing>
+		{#if $message}
+			<label class="label">
+				<span class="label-text-alt text-lg text-error mb-2">{$message}</span>
+			</label>
+		{/if}
+		<button
+			on:click={() => modal.submitModal()}
+			class="btn btn-primary"
+			class:loading={closing}>
```

Comment:
`loading={submitting}` might be easier to read for some... `let` on that fragment like that, while is valid, tends to be a rarely used technique since it's such a rare situation to do it.  The typical approach for all of this would for the `Form` to dispatch a `submitting` event that can be utilized by this page to turn he `loading` class on/off.

Reply:
alright I'll have to think about that. I think it's still useful in some cases (like the delete dialog, there's not always a form there), but having something on the form would make a lot of sense.
That being said a lot of times with a dialog you're doing 2 things, you submit the data, but then you may also need to refresh the local data to reflect those changes, and ideally the dialog stays up while the new data is loading, I'm not really sure how that would fit with a form submitting as I'm not sure when the submitting might change back to `false`.

Reply:
some of this is similar to my response in https://github.com/sillsdev/languageforge-lexbox/pull/30#discussion_r1154749333